contracts/src/v0.8/dev/VRF.sol

```diff
@@ -0,0 +1,537 @@
+        bigModExpContractInputs[2] = WORD_LENGTH_BYTES;  // Length of modulus
+        bigModExpContractInputs[3] = base;
+        bigModExpContractInputs[4] = exponent;
+        bigModExpContractInputs[5] = FIELD_SIZE;
+        uint256[1] memory output;
+        assembly { // solhint-disable-line no-inline-assembly
+            callResult := staticcall(
+            not(0),                   // Gas cost: no limit
+            0x05,                     // Bigmodexp contract address
+            bigModExpContractInputs,
+            0xc0,                     // Length of input segment: 6*0x20-bytes
+            output,
+            0x20                      // Length of output segment
+            )
+        }
+        if (callResult == 0) {revert("bigModExp failure!");}
+        return output[0];
+    }
+
+    // Let q=FIELD_SIZE. q % 4 = 3, ∴ x≡r^2 mod q ⇒ x^SQRT_POWER≡±r mod q.  See
+    // https://en.wikipedia.org/wiki/Modular_square_root#Prime_or_prime_power_modulus
+    uint256 constant private SQRT_POWER = (FIELD_SIZE + 1) >> 2;
+
+    // Computes a s.t. a^2 = x in the field. Assumes a exists
+    function squareRoot(uint256 x) internal view returns (uint256) {
+        return bigModExp(x, SQRT_POWER);
+    }
+
+    // The value of y^2 given that (x,y) is on secp256k1.
+    function ySquared(uint256 x) internal pure returns (uint256) {
+        // Curve is y^2=x^3+7. See section 2.4.1 of https://www.secg.org/sec2-v2.pdf
+        uint256 xCubed = mulmod(x, mulmod(x, x, FIELD_SIZE), FIELD_SIZE);
+        return addmod(xCubed, 7, FIELD_SIZE);
+    }
+
+    // True iff p is on secp256k1
+    function isOnCurve(uint256[2] memory p) internal pure returns (bool) {
+        return ySquared(p[0]) == mulmod(p[1], p[1], FIELD_SIZE);
+    }
+
+    // Hash x uniformly into {0, ..., FIELD_SIZE-1}.
+    function fieldHash(bytes memory b) internal pure returns (uint256 x_) {
+        x_ = uint256(keccak256(b));
+        // Rejecting if x >= FIELD_SIZE corresponds to step 2.1 in section 2.3.4 of
+        // http://www.secg.org/sec1-v2.pdf , which is part of the definition of
+        // string_to_point in the IETF draft
+        while (x_ >= FIELD_SIZE) {
+            x_ = uint256(keccak256(abi.encodePacked(x_)));
+        }
+    }
+
+    // Hash b to a random point which hopefully lies on secp256k1. The y ordinate
+    // is always even, due to
+    // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vrf-05#section-5.4.1.1
+    // step 5.C, which references arbitrary_string_to_point, defined in
+    // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vrf-05#section-5.5 as
+    // returning the point with given x ordinate, and even y ordinate.
+    function newCandidateSecp256k1Point(bytes memory b)
+    internal view returns (uint256[2] memory p) {
+        p[0] = fieldHash(b);
+        p[1] = squareRoot(ySquared(p[0]));
+        if (p[1] % 2 == 1) {
+            p[1] = FIELD_SIZE - p[1];
```

Comment:
added a comment explaining why thats not needed

Reply:
I guess what I'm trying to say is that it would be preferable to just wrap the entire bodies of all these functions in `unchecked`. That way we don't deviate from the audited semantics AND we save gas/bytecode because we don't do unnecessary overflow checks.

Reply:
Good point

Reply:
Done